### ISSUE TITLE
allow the @ServiceName injection to use http or https as the protocol depending on the service port (80 / 443)?

### ISSUE THREAD

**Issue by jstrachan** [fabric8io/fabric8#3808]
so gogs can be configured to use http or https; we've code that would like to use whichever flavour is being used.

https://github.com/fabric8io/fabric8/blob/master/forge/fabric8-forge-core/src/main/java/io/fabric8/forge/rest/main/GitUserHelper.java#L44-44

I added @Protocol("http") as otherwise this results in URLs of the form: "tcp://foo.bar" which don't work too well in browsers.

What would be nice is if we know its either http or https then use one of those protocols and pick the right protocol based on the port number (e.g. if 443 then use https otherwise default to http).

I kinda hoped the env vars could expose the protocol; but seems to always be "TCP". e.g.:
```
            "GOGS_HTTP_SERVICE_PORT_443_TCP_PROTO=tcp",
```

Ah - looks like we're not setting the protocol string on the gogs service; lets try that and see if it helps...

**Comment by jstrachan** [comment]
am patching the mvn plugin to allow <code>fabric8.service.protocol</code> to be used...

**Comment by jimmidyson** [comment]
Service protocol can only be tcp or UDP AFAIK. This is for the kube proxy
to set up port listeners.

**Comment by jstrachan** [comment]
ah crap! :(

**Comment by jstrachan** [comment]
OK maybe we need a @HttpProtocol annotation for the CDI stuff; to pick the most suitable http protocol for injecting URLs?

**Comment by jimmidyson** [comment]
Perhaps a bit more generic, can imagine ssh, ftp, git, etc all being
options? Defaulting to http seems sane though.

**Comment by jstrachan** [comment]
Yeah, maybe a @ProtocolFromPort that uses the actual port numbers of the service to determine http/https/ssh etc?

**Comment by jstrachan** [comment]
If a service exposes ssh and http this could be confusing ;). Maybe we need to use a @HttpProtocol annotation to pick https if possible or http if not for the protocol?

**Comment by jstrachan** [comment]
For specific protocols, like ssh, folks can use <code>@Protocol("ssh")</code> its only for things like http/https that a service may or may not be configured to use https; so if not fall back to http (e.g. for gogs).

**Comment by iocanel** [comment]
I think that we should just add an option to the existing annotation Protocol annotation and if explicitly set, we should attempt a "port to protocol" conversion.
